### ISSUE TITLE
How to deal with augmentations which exceed max/min amplitude?

### ISSUE THREAD

**Issue by bml1g12** [iver56/audiomentations#25]
Thanks for the great library! I noticed some of the transforms seem to expand the volume beyond +1 and -1; how do you suggest dealing with transforms that do this? If we scale the entire volume down to fit, then in some cases you could end up accidently making the mean volume too small to be audible.

A related issue is that data is in 16 bit signed integers (using PyDub library) so for this library is it correct I need to convert to 32 bit floats in the range -1 to +1? In going back and forth between these I am using the following functions, but I'd be grateful for any advice if this is incorrect:

```
def int16_samples_to_float32(y):
    """Convert int16 numpy array of audio samples to float32."""
    if y.dtype != np.int16:
        raise ValueError('input samples not int16')
    return y.astype(np.float32) / np.iinfo(np.int16).max

def float_samples_to_int16(y):
    """Convert floating-point numpy array of audio samples to int16."""
    if not issubclass(y.dtype.type, np.floating):
        raise ValueError('input samples not floating-point')
    try:
        result = (y * np.iinfo(np.int16).max).astype(np.int16)
    except Warning:
        print(y, np.iinfo(np.int16).max) # sometimes catch warnings that limit has been exceeded
        raise
    return result
 ```

**Comment by iver56** [comment]
Interesting! Do you have a list of transforms that expand the max amplitude beyond +1 and -1? Maybe we should add those cases as unit tests so we can study the cases? Help is appreciated :)

Maybe we could add a "limiter" transform or "compressor" transform? This is what is usually done by mobile recorders and music producers (in DAWs) when they need to gracefully* reduce too loud peaks in their signal. *Bring down the volume without adding too much unwanted distortion.

Yes, `audiomentations` does indeed expect the audio samples to be float32 and not int16. And yes, the range should be -1 to +1. Maybe you could compare your functions with open source audio processing/utility libraries? Or see what other people do in their code? A quick search returns these resources:

* https://github.com/tensorflow/magenta/blob/master/magenta/music/audio_io.py#L41
* https://stackoverflow.com/a/42544738/2319697

I note from that Stackoverflow conversation that samples in int16 should not have value -32768. Hence, it may be safer to multiply by 32767 instead of 32768, to avoid integer underflow

**Comment by bml1g12** [comment]
Sorry for the delay - haven't had any time to work on this since I posted the issue. Yes AddGuassianNoise definately does, and I think possibly also FrequencyMask (at least I seemed to end up with NaN or inf with that). I was using a set of them when I encountered this issue - so to figure out which is causing the issue would require some more extensive testing in isolation. 

Yeah maybe a solution but if it would need to be done before the relevant transform (e.g. AddGaussianNoise) in such a way as to guarantee the latter transform doesn't exceed the max - so maybe a way of coupling it in a pair with existing transforms - sounds a little complicated though...Another way might to train a min-max scaler on the input to a transform, then scale the output of the transform to match the initial input (? not sure if this would cause unintended side effects).  

I've noticed there is a surprisingly small amount of literature on Audio augmentation for machine learning - on kaggle most people seem to do simple stretch and volume augmentation - so I think this library is pretty cool. 

Oh thanks for the helpful links - I'm using the first link's suggestion at the moment.
